mio/src/lib.rs

```diff
@@ -1,111 +1,145 @@
 extern crate mio;
 extern crate futures;
+extern crate fnv;
 
+use std::hash::BuildHasherDefault;
 use std::collections::HashMap;
 use std::io::{self, ErrorKind, Read, Write};
 use std::net::SocketAddr;
 use std::panic;
 use std::slice;
 use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::mpsc::{channel, TryRecvError};
 
-use futures::{Future, promise, Complete, PollError};
+use fnv::FnvHasher;
 
-pub type IoFuture<T> = Future<Item=T, Error=io::Error>;
+use futures::{Future, Tokens, promise, Complete, Wake, PollError, PollResult};
 
-pub struct Loop {
-    io: mio::Poll,
-    tx: mio::channel::Sender<Message>,
-    rx: mio::channel::Receiver<Message>,
-    next: usize,
-    done: HashMap<usize, Complete<(), io::Error>>,
+pub struct MioEvent {
+    events: mio::EventSet,
+    source: Source,
+    loop_handle: LoopHandle,
 }
 
-enum Message {
-    Wait(Complete<(), io::Error>, mio::EventSet, Arc<mio::Evented + Send + Sync>),
-    Register(Arc<mio::Evented + Send + Sync>),
+impl MioEvent {
+    fn into_future<C: PollCompletion>(self, completion: C) -> MioFuture<C> {
+        MioFuture {
+            event: self,
+            completion: completion,
+        }
+    }
+}
+
+pub trait PollCompletion {
+    type Item: Send + 'static;
+    type Error: Send + 'static;
+
+    fn poll_completion(&mut self) -> Option<PollResult<Self::Item, Self::Error>>;
 }
 
+pub struct MioFuture<C> {
+    event: MioEvent,
+    completion: C,
+}
```

Comment:
I think it definitely makes sense to have a central bundle of:

* What we're waiting on
* What we're waiting for
* How we know when we're done

and all that packaged up is a future